benchmark/single-source/FloatingPointParsing.swift

```diff
@@ -0,0 +1,273 @@
+//===--- FloatingPointParsing.swift -----------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// This test verifies the performance of parsing a floating point value
+// from a String.
+
+import TestsUtils
+
+public let FloatingPointParsing = [
+  BenchmarkInfo(
+    name: "ParseFloat.Float.Exp",
+    runFunction: run_ParseFloatExp,
```

Comment:
Since you did the nice generic workload function, you can do this, too:
```suggestion
    runFunction: { parseFloatWorkload($0, workload: floatExpWorkload, type: Float.self) },
```

Reply:
There's a slight semantic difference there: you lose the `@inline(never)`. I'm not sure that matters, but it is a difference – and you can't make up for it by putting it on the generic function since that would cause it to be unspecialised.

Also, you could only do clean one-line closures for `Float` and `Double`, since `Float80` needs to be conditionally excluded; at that point, it makes sense to make the `Float80` variants a separate method, and if you're doing that I think you may as well have separate methods for all of them. I'm not strongly tied to that, though.

Reply:
You can put @inline(never)before `parseFloatWorkload` if you wish, but it doesn't really matter in the end. That syntactic tradition seems to be a remnant of the time before introduction of `BenchmarkInfo`.

Regarding the platform #if/else compilation directives, I think it would be cleaner to wrap them around the `BenchmarkInfo` declarations for `Float80` directly, anyway...

Reply:
> You can put @inline(never) before parseFloatWorkload if you wish, but it doesn't really matter in the end. 

No, he wants `@inline(__always)` for more guaranteed specialization, otherwise we're measuring a very different thing.

> Also I don't think specialization is related to inlining as you suggest, see DataBenchmarks.swift...

Unfortunately, specialization is *very* related to inlining and it will be a long time before this is improved. `@inline(never)`, as in DataBenchmarks.swift, is likely to suppress specialization which is very useful for benchmarking hand-coded specializations inside the API (and Data has many).

I believe @troughton's formulation is ideal.

Reply:
@milesman In DataBenchmarks, I have tried both formulations before settling on the `@inline(never)` on the generic function like `append`… are you saying it is incorrect?

Reply:
@milseman In DataBenchmarks, I have tried both formulations before settling on the `@inline(never)` on the generic function like `append`… are you saying it might be incorrect?